tests/core/test_optimizable.py

```diff
@@ -1324,6 +1328,13 @@ def test_adder_dofs_shared_fix_unfix(self):
 
         adder_orig.fix_all()
         self.assertTrue(len(adder_shared_dofs.x) == 0)
+    
+    def test_derivative(self):
+        adder_orig = OptClassSharedDOFs(x0=[1, 2, 3], names=["x", "y", "z"],
+                                        fixed=[False, False, True])
+        adder_shared_dofs = OptClassSharedDOFs(dofs=adder_orig.dofs)
+        sum_obj = adder_orig + adder_shared_dofs
+        self.assertEqual(adder_orig.dJ()*2, sum_obj.dJ())
```

Comment:
@mbkumar the `Optimizable` framework is not combining derivatives correctly for `Optimizable`s that share `DOF`s, see above example.

Reply:
This is not the deficiency of the Optimizable framework. The Derivative class uses the Optimizable objects as keys. Since each instance is a different object, the keys are different. So Derivative class is giving a 4 element vector for sum_obj.dJ(). 

IMO, the Derivative class needs to be updated to account for DOF sharing.

Reply:
I have attempted a fix in the `__call__` dunder

Reply:
@andrewgiuliani, @landreman, @florianwechsung 

I think this issue is more general than just derivatives. What if the final objective contains terms from two or more Optimizable objects that share a single DOFs object. The global state vector contains duplicates of the dofs from the shared DOFs object. We need to have a more rigorous framework for assembling the global state vector.
My question is will this situation arise or is the sharing of DOFs limited to simple cases?

Reply:
I updated the global dof vector assembly to account for shared DOFs.